test/e2e/framework/util.go

```diff
@@ -0,0 +1,85 @@
+package framework
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+// ArtifactDir determines where artifacts should live for a test case.
+func ArtifactDir(t TestingTInterface, segments ...string) (string, error) {
+	var baseDir string
+	if dir, set := os.LookupEnv("ARTIFACT_DIR"); set {
+		baseDir = dir
+	} else {
+		baseDir = t.TempDir()
+	}
+	segments = append([]string{
+		baseDir, strings.NewReplacer("/", "_", "\\", "_", ":", "_").Replace(t.Name()),
+	}, segments...)
+	artifactDir := filepath.Join(segments...)
+	if err := os.MkdirAll(artifactDir, 0755); err != nil {
+		return "", fmt.Errorf("could not create artifact dir: %v", err)
+	}
+	t.Logf("Saving artifacts to %s.", artifactDir)
+	return artifactDir, nil
+}
+
+// GetFreePort asks the kernel for a free open port that is ready to use.
+func GetFreePort(t TestingTInterface) (string, error) {
+	for {
+		addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+		if err != nil {
+			return "", fmt.Errorf("could not resolve free port: %w", err)
+		}
+
+		l, err := net.ListenTCP("tcp", addr)
+		if err != nil {
+			return "", fmt.Errorf("could not listen on free port: %w", err)
+		}
+		defer func(c io.Closer) {
+			if err := c.Close(); err != nil {
+				t.Errorf("could not close listener: %v", err)
+			}
+		}(l)
+		port := strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
+		// Tests run in -parallel will run in separate processes, so we must use the file-system
+		// for sharing state and locking across them to coordinate who gets which port. Without
+		// some mechanism for sharing state, the following race is possible:
+		// - process A calls net.ListenTCP() to resolve a new port
+		// - process A calls l.Close() to close the listener to allow accessory to use it
+		// - process B calls net.ListenTCP() and resolves the same port
+		// - process A attempts to use the port, fails as it is in use
+		// Therefore, holding the listener open is our kernel-based lock for this process, and while
+		// we hold it open we must record our intent to disk.
+		lockDir := filepath.Join(os.TempDir(), "kcp-e2e-ports")
+		if err := os.MkdirAll(lockDir, 0777); err != nil {
+			return "", fmt.Errorf("could not create port lockfile dir: %v", err)
+		}
+		lockFile := filepath.Join(lockDir, port)
+		if _, err := os.Stat(lockFile); os.IsNotExist(err) {
+			// we've never seen this before, we can use it
+			f, err := os.Create(lockFile)
+			if err != nil {
+				return "", fmt.Errorf("could not record port lockfile: %w", err)
```

Comment:
No, because we're using the kernel-provided connection open as a "lock" for this critical section. Nobody else can listen on that port before we're done.